### PR TITLE
Refactor and clean up code to follow Sklearn convention

### PR DESCRIPTION
**Pull request by kmzzhang**
- save all parameters and hyperparameters (x, y scales, and network hyperparams) to state dict for simplicity
- Use state_dict keyword in place of featurizer and flow keywords for initializing model
  -  this makes it easier to distribute model now that network hyperparams are saved to the state_dict
- use set_params, get_params, save_params method names
- other refactoring

### CODE REVIEW

File: src/nbi/engine.py [hunk 1]
```diff
@@ -703,79 +672,70 @@ def get_network(self):
...
+        }
+        return state_dict
 
-    def save_state_dict(self):
+    def save_params(self, path):
         """
         Saves the network weights and pre-processing scales to disk
 
         Returns
         -------
```
Comment (stefanv) [inline src/nbi/engine.py:735]:
```suggestion
```

**Comment by kmzzhang** [comment]
@stefanv I expanded integration tests to include the new features. Request merge

File: test/test_engine.py [hunk 2]
```diff
@@ -1,3 +1,8 @@
+import warnings
+
+warnings.filterwarnings("ignore")
+warnings.simplefilter("ignore")
```
[the remark this answers is not in the document]
Comment (kmzzhang) [reply test/test_engine.py:4]: I did it for my own debugging. Could remove it needed

File: src/nbi/engine.py (hunk 1 above)
Comment (kmzzhang) [reply src/nbi/engine.py]: Are you suggesting that all functions without returns should not have "Returns" in the doctoring?

File: src/nbi/engine.py [hunk 3]
```diff
@@ -395,64 +408,42 @@ def fit(
             if self.round > 0 and plot:
                 self.weighted_corner(x_obs, y_true)
 
-            # early stopping
+            # stop further rounds of SNPE?
+            # option 1: enough effective posterior samples
             if np.sum(self.neff) > neff_stop > 0:
                 print("Success: Exceed specified stopping sample size!")
```
[the remark this answers is not in the document]
Comment (kmzzhang) [reply src/nbi/engine.py:438]: Let's fix this in a separate PR

File: src/nbi/engine.py [hunk 4]
```diff
@@ -369,24 +374,32 @@ def fit(
...
+                        self.directory, str(self.round), str(epoch_best) + ".pth"
                     )
-                    self.load_state_dict(self.epoch - early_stop_patience - 2)
+                    self.set_params(path)
                     break
 
-            self.save_current_state()
             self.round += 1
 
+            # stop here for ANPE
```
[the remark this answers is not in the document]
Comment (kmzzhang) [reply src/nbi/engine.py:426]: Amortized Neural Posterior Estimation. I will clarify in the comments

File: src/nbi/engine.py [hunk 5]
```diff
@@ -369,24 +374,32 @@ def fit(
...
+                if epoch_best < (epoch - early_stop_patience):
                     print(
                         "early stopping, loading state dict from epoch",
-                        self.epoch - early_stop_patience - 2,
+                        epoch_best,
+                    )
+
+                    # load from the epoch with lowest validation loss
+                    path = os.path.join(
+                        self.directory, str(self.round), str(epoch_best) + ".pth"
```
[the remark this answers is not in the document]
Comment (kmzzhang) [reply src/nbi/engine.py:418]: Let's fix this in a seperate PR

File: src/nbi/engine.py [hunk 6]
```diff
@@ -369,24 +374,32 @@ def fit(
...
                     )
 
-                self.save_state_dict()
+                self.save_params(
+                    os.path.join(self.directory, str(self.round), str(epoch) + ".pth")
+                )
 
-                if self.stop_training(early_stop_patience):
+                epoch_best = np.argmin(self.val_losses[-1])
+                if epoch_best < (epoch - early_stop_patience):
```
[the remark this answers is not in the document]
Comment (kmzzhang) [reply src/nbi/engine.py:410]: fixed

File: src/nbi/engine.py [hunk 7]
```diff
@@ -369,24 +374,32 @@ def fit(
...
                         {
-                            "Train Loss": self.training_losses[-1],
-                            "Val Loss": self.validation_losses[-1],
+                            "Train Loss": self.train_losses[-1][-1],
+                            "Val Loss": self.val_losses[-1][-1],
                         }
                     )
 
-                self.save_state_dict()
+                self.save_params(
```
[the remark this answers is not in the document]
Comment (kmzzhang) [reply src/nbi/engine.py]: The ideas is that users might want to inspect the model history over rounds and epochs

File: src/nbi/engine.py [hunk 8]
```diff
@@ -369,24 +374,32 @@ def fit(
                 if self.wandb:
```
[the remark this answers is not in the document]
Comment (kmzzhang) [reply src/nbi/engine.py:395]: I will fix all wandb related issues in a separate PR. Wandb is currently non functional

File: src/nbi/__init__.py [hunk 9]
```diff
@@ -7,6 +7,6 @@
 __url__ = "https://github.com/kmzzhang/nbi"
 __license__ = "BSD-3"
 __description__ = "Neural Bayesian Inference"
-__version__ = "0.4.1"
+__version__ = "0.5.1"
```
[the remark this answers is not in the document]
Comment (kmzzhang) [reply src/nbi/__init__.py]: Reverted

**Pull request merged**